api/v1alpha1/helpers.go

```diff
@@ -0,0 +1,62 @@
+/*
+ Copyright 2024, NVIDIA CORPORATION & AFFILIATES
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+     http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"math/big"
+	"net"
+
+	"github.com/Mellanox/nvidia-k8s-ipam/pkg/ip"
+)
+
+// GetGatewayForSubnet returns computed gateway for subnet as string, index 0 means no gw
+// Examples:
+// - subnet 192.168.0.0/24, index 1, gateway = 192.168.1.1 (skip network IP)
+// - subnet 192.168.1.2/31, index 1, gateway = 192.168.1.2 (use network IPs for point to point prefixes /31, /127)
+// - subnet 192.168.0.0/24, index 10, gateway = 102.168.1.10
+// - subnet 192.168.33.0/24, index 900, gateway = "" (invalid config, index too large)
+// - subnet 192.168.33.0/24, index 0, gateway = "" (no gw)
+func GetGatewayForSubnet(subnet *net.IPNet, index uint) string {
+	if index == 0 {
+		return ""
+	}
+	if getPossibleIPCount(subnet).Cmp(big.NewInt(int64(index))) < 0 {
+		// index too large
+		return ""
+	}
+	setBits, bitsTotal := subnet.Mask.Size()
+	if setBits == bitsTotal-1 {
+		// point to point prefix
+		index--
+	}
+	gwIP := ip.NextIPWithOffset(subnet.IP, int64(index))
+	if gwIP == nil {
+		return ""
+	}
+	return gwIP.String()
+}
+
+func getPossibleIPCount(subnet *net.IPNet) *big.Int {
+	setBits, bitsTotal := subnet.Mask.Size()
+	isPointToPoint := setBits == bitsTotal-1
+	ipCount := big.NewInt(0).Exp(big.NewInt(2), big.NewInt(int64(bitsTotal-setBits)), nil)
+	if !isPointToPoint {
+		if subnet.IP.To4() != nil {
+			ipCount.Sub(ipCount, big.NewInt(2))
+		} else {
+			ipCount.Sub(ipCount, big.NewInt(1))
+		}
```

Comment:
Looks like this is a new condition right? We didn't have this before in `IPPool` validation. 

Maybe worth adding tests for that bit to show the behavior diff between ipv4/ipv6

Reply:
done